lib/SIL/Notifications.cpp

```diff
@@ -0,0 +1,54 @@
+//===--- Notifications.cpp ------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-notifications"
+#include "swift/SIL/Notifications.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace swift;
+
+//===----------------------------------------------------------------------===//
+// DeserializationNotificationHandler Impl of
+// DeserializationNotificationHandlerSet
+//===----------------------------------------------------------------------===//
+
+#ifdef DNS_CHAIN_METHOD
+#error "DNS_CHAIN_METHOD is defined?!"
+#endif
+#define DNS_CHAIN_METHOD(Name, FirstTy, SecondTy)                              \
+  void DeserializationNotificationHandlerSet::did##Name(FirstTy first,         \
+                                                        SecondTy second) {     \
+    LLVM_DEBUG(llvm::dbgs()                                                    \
+               << "*** Deserialization Notification: " << #Name << " ***\n");  \
+    for (auto *p : getRange()) {                                               \
+      LLVM_DEBUG(llvm::dbgs()                                                  \
+                 << "    Begin Notifying: " << p->getName() << "\n");          \
+      p->did##Name(first, second);                                             \
+      LLVM_DEBUG(llvm::dbgs()                                                  \
+                 << "    End Notifying: " << p->getName() << "\n");            \
+    }                                                                          \
+    LLVM_DEBUG(llvm::dbgs()                                                    \
+               << "*** Completed Deserialization Notifications for " #Name     \
+                  "\n");                                                       \
+  }
```

Comment:
Here the net benefit of macros is obvious, but a bit sad. I don't have a much better suggestion.

If you had a single DO_DNS_CHAIN_METHOD macro instead that specified the entire interface, then using the same macro everywhere would actually make sense. I don't think using a macro everywhere is a good goal though.

Reply:
I will just use the macro here and eliminate the one from the header. Originally they were both in the header, but I did not want to add the logging stuff to the header.